### PR TITLE
[SYCL][thinLTO][ClangLinkerWrapper] Add option for thinLTO

### PR DESCRIPTION
**Pull request by sarnex**
Add an option for thinLTO with sycl to `clang-linker-wrapper` and start to use it.

It's technically possible to determine if a BC file was compiled with thinLTO, but given that this mode being enabled has a significant effect on our tools process and it would be difficult to rework the code to first check if the BC files are thinLTO and change the tools based on that, an option seems simpler and allows us to catch mismatches with the actual BC files given and error.

### CODE REVIEW

**Comment by asudarsa** [comment]
Hi @sarnex 

One quick question. Is it possible to send a mixture of thin-LTO and non thin-LTO compiled bc files to the linker-wrapper?
If that is a possibility, it seems like this can be supported in the StringMap of the Image rather than sending it as an aotion.
If the mix is not possible, then the option sounds good.

Thanks

**Comment by sarnex** [comment]
@asudarsa No, it won't work and I will add an error for that case using this option

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp [hunk 1]
```diff
@@ -570,10 +570,10 @@ getTripleBasedSYCLPostLinkOpts(const ArgList &Args,
   bool SplitEsimd =
       Args.hasFlag(OPT_sycl_device_code_split_esimd,
                    OPT_no_sycl_device_code_split_esimd, SplitEsimdByDefault);
-
-  // Symbol file and specialization constant info generation is mandatory -
+  if (!Args.hasArg(OPT_sycl_thin_lto))
+    PostLinkArgs.push_back("-symbols");
+  // Specialization constant info generation is mandatory -
```
Comment (asudarsa) [inline clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:575]: I do not know how and when this comment got here, I can handle this in a future commit.

Thanks

File: clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td [hunk 2]
```diff
@@ -185,3 +185,6 @@ def sycl_add_default_spec_consts_image : Flag<["--", "-"], "sycl-add-default-spe
   Flags<[WrapperOnlyOption, HelpHidden]>;
 def no_sycl_add_default_spec_consts_image : Flag<["--", "-"], "no-sycl-add-default-spec-consts-image">,
   Flags<[WrapperOnlyOption, HelpHidden]>;
+
+def sycl_thin_lto : Flag<["--", "-"], "sycl-thin-lto">,
+Flags<[WrapperOnlyOption]>,  HelpText<"Link SYCL device code using thinLTO">;
```
Comment (asudarsa) [inline clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td:190]: EOF please

**Review by asudarsa — APPROVED** [review]
LGTM. Thanks for incremental PR. :-)

**Comment by sarnex** [comment]
@asudarsa Thanks, yeah I'm trying to do smaller PRs rather than drop a huge PR that is impossible to review

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp (hunk 1 above)
Comment (sarnex) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:575]: sure, thx

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 3]
```diff
@@ -11149,6 +11149,11 @@ void LinkerWrapper::ConstructJob(Compilation &C, const JobAction &JA,
       appendOption(OptString, A);
     CmdArgs.push_back(Args.MakeArgString("--llvm-spirv-options=" + OptString));
 
+    bool IsUsingLTO = D.isUsingLTO(/*IsDeviceOffloadAction=*/true);
+    auto LTOMode = D.getLTOMode(/*IsDeviceOffloadAction=*/true);
+    if (IsUsingLTO && LTOMode == LTOK_Thin)
+      CmdArgs.push_back(Args.MakeArgString("-sycl-thin-lto"));
```
[the remark this answers is not in the document]
Comment (sarnex) [reply clang/lib/Driver/ToolChains/Clang.cpp:11155]: agree, thx

File: clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td (hunk 2 above)
Comment (sarnex) [reply clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td]: done in latest commit, thank you!

**Review by mdtoguchi — APPROVED** [review]

**Pull request merged**